dtm0/service.c

```diff
@@ -0,0 +1,648 @@
+	       " curr_proc_fid="FID_F,
+	       evented_proc_state,
+	       FID_P(evented_proc_fid),
+	       FID_P(&process->dop_rserv_fid),
+	       FID_P(&process->dop_rproc_fid),
+	       FID_P(current_proc_fid));
+	M0_LOG(M0_DEBUG, "dtm0=%p dop_rep=%s", dtm0, process->dop_rep);
+
+	switch (evented_proc_state) {
+	case M0_NC_ONLINE:
+		process->dop_service_connect_ast = (struct m0_sm_ast){
+			.sa_cb    = &service_connect_ast,
+			.sa_datum = process,
+		};
+		m0_sm_ast_post(m0_locality_here()->lo_grp,
+			       &process->dop_service_connect_ast);
+		break;
+	case M0_NC_TRANSIENT:
+		ready_to_process_ha_msgs = 1;
+		M0_LOG(M0_DEBUG, "TRANSIENT received, now is "
+		       "ready to process HA messages");
+		break;
+	default:
+		M0_LOG(M0_DEBUG, "Ignored received proc state %d",
+		       evented_proc_state);
+		break;
+	}
+out:
+	M0_LEAVE();
+
+	return false;
+}
+
+static void dtm0_process__ha_state_subscribe(struct dtm0_process *process,
+					     struct m0_conf_obj  *obj)
+{
+	M0_ENTRY();
+	m0_clink_init(&process->dop_ha_link, process_clink_cb);
+	m0_clink_add_lock(&obj->co_ha_chan, &process->dop_ha_link);
+	M0_LEAVE();
+}
+
+static void dtm0_process__ha_state_unsubscribe(struct dtm0_process *process)
+{
+	M0_ENTRY();
+	m0_clink_del_lock(&process->dop_ha_link);
+	m0_clink_fini(&process->dop_ha_link);
+	M0_LEAVE();
+}
+
+static bool conf_obj_is_process(const struct m0_conf_obj *obj)
+{
+	return m0_conf_obj_type(obj) == &M0_CONF_PROCESS_TYPE;
+}
+
+static int dtm0_service__ha_subscribe(struct m0_reqh_service *service)
+{
+	struct m0_confc        *confc = m0_reqh2confc(service->rs_reqh);
+	struct m0_conf_root    *root;
+	struct m0_conf_diter    it;
+	struct m0_conf_obj     *obj;
+	struct m0_conf_process *process;
+	struct dtm0_process    *dtm0_process;
+	struct m0_dtm0_service *s;
+	struct m0_fid           rproc_fid;
+	struct m0_fid           rserv_fid;
+	struct m0_fid          *current_proc_fid = &service->rs_reqh->rh_fid;
+
+	int rc;
+
+	M0_ENTRY();
+
+	M0_PRE(service != NULL);
+	s = container_of(service, struct m0_dtm0_service, dos_generic);
+
+	/** UT workaround */
+	if (!m0_confc_is_inited(confc)) {
+		M0_LOG(M0_WARN, "confc is not initiated!");
+		return M0_RC(0);
+	}
+
+	rc = m0_confc_root_open(confc, &root);
+	if (rc != 0)
+		return M0_ERR(rc);
+
+	rc = m0_conf_diter_init(&it, confc,
+				&root->rt_obj,
+				M0_CONF_ROOT_NODES_FID,
+				M0_CONF_NODE_PROCESSES_FID);
+	if (rc != 0) {
+		m0_confc_close(&root->rt_obj);
+		return M0_ERR(rc);
+	}
+
+	while ((rc = m0_conf_diter_next_sync(&it, conf_obj_is_process)) > 0) {
+		obj = m0_conf_diter_result(&it);
+		process = M0_CONF_CAST(obj, m0_conf_process);
+		rproc_fid = process->pc_obj.co_id;
+		rc = m0_conf_process2service_get(confc, &rproc_fid,
+						 M0_CST_DTM0, &rserv_fid);
+		if (rc == 0) {
+			if (m0_fid_eq(&rproc_fid, current_proc_fid))
+				/* skip current process */
+				continue;
+			M0_ALLOC_PTR(dtm0_process);
+			M0_ASSERT(dtm0_process != NULL); /* XXX */
```

Comment:
error handling

Reply:
Will be fixed during DTM0 start/stop implementation, behavior may change.